### ISSUE TITLE
[验证] DeepSeek-V2-Lite Q4 路径与 numpy 参考不一致（#104 遗留）

### ISSUE THREAD

**Issue by gqf2008** [gqf2008/machserve#107]
[验证] DeepSeek-V2-Lite Q4 路径与 numpy 参考不一致（#104 遗留）

## 背景

#104 定位并修复了 RoPE 配对约定错误（DeepSeek 用相邻对，此前统一按半半配对
实现）。该修复已被两个独立实现数值证实，见 PR（RoPE 提交）。

但 DeepSeek-V2-Lite-Chat 在 mach-server 上的端到端输出**仍不连贯**，与
numpy 参考不一致。这一项未能定位根因，按 #104 验收标准另立本 issue。

## 现象

prompt `"The capital of France is"`，temperature 0，max_tokens 64。

- numpy 参考（bf16 权重，f32/f64 计算）：`' Paris.'`（logit 33.22，
  领先次选 `' The'` 31.78 共 1.44）
- numpy 参考 + **模拟 Q4**（见下）：`' The capital of France is Paris.'`
  （step 0 的 top-1 变成 `' The'` 30.71 / `' Paris'` 30.04，但整句连贯，
  step 7 正常 EOS）
- mach-server（Q4-on-device）：`' Languages Capital Languages Language...'`
  首 token 是 60750（`' Languages'`），随后退化为 31939 的重复

60750 在 numpy 的 top-20 里完全不出现（top-20 的下限是 22.85），即服务端
的 argmax 比 numpy 的 top-1 高出约 7 个 logit 量级的偏离。

## 已排除的假设

**1. Q4 量化质量不足 —— 已排除（存疑，见下）**

在 numpy 参考里复现了与 `q4.rs` 完全一致的量化（Q4_GROUP=32，
`scale = max|w|/7`，对称 int4，clamp 到 [-7,7]），并且按 loader 的实际行为
量化**所有** GEMM 权重（attention + dense MLP + 路由专家 + shared experts +
embed_tokens + lm_head），norm/bias 保持 f32。

结果：模型仍然连贯，输出 `' The capital of France is Paris.'` 并在 step 7
正常停。也就是说 int4 本身不足以毁掉这个 checkpoint。

注：我的模拟比真实的 Q4 路径**更苛刻** —— Rust loader 把 MoE router
（`mlp.gate.weight`）留在 f32 不量化（loader.rs:889），而我的模拟把它也量化
了；但即便如此参考依然连贯。

**2. Q4 核内反量化 kernel 有 bug —— 已排除**

`moe_grouped_gate_up_q4` / `moe_grouped_down_q4`（kernels.rs:2118 / 2227）
是 #85 引入的"专家池留在设备端、核内反量化"路径，此前**没有任何测试覆盖**。

用 8 层截断 checkpoint 做 A/B：同一份权重，只切换 `MACH_Q4_DEVICE`。

| 配置 | 首 token |
|---|---|
| f16（无 Q4） | 43666 |
| Q4 + `MACH_Q4_DEVICE=1`（核内反量化） | 11763 |
| Q4 无 `MACH_Q4_DEVICE`（host 反量化后上传 f16） | 11763 |

两个 Q4 路径给出**完全相同的首 token**，说明核内反量化与 host 反量化一致，
不是元凶。

**3. f16 前向路径有 bug —— 已排除**

同一 8 层截断模型，f16 服务端的首 token 是 43666，恰好是 numpy NLAY=8 的
**次选**（numpy top-1 是 10244 17.66，次选 43666 17.48，间距仅 0.18）。
0.18 的间距落在 f16 舍入误差内，说明 f16 路径基本正确。

**4. 缓冲区越界 —— 已排除**

`batched.rs:2255` 的 `launch_rope_batched` 传 `n_kv_heads = heads = 16`、
`head_dim = rope = 64`，看起来会往 `k_buf` 写 `b*16*64` 个 float。但
`k_buf` 的分配是 `b * nkv * 4` 而 `nkv = n_kv_heads * c.head_dim = 16*192`
（batched.rs:1071/1103），即 `b*3072` 个 float，远大于所需的 `b*1024`。
不越界。

**5. 其他已核对无误的点**

- `mla_assemble_kv_batched` 的 k_nope/v 切分顺序与 numpy 一致
  （`k_nope = kv[h][:nope]`，`v = kv[h][nope:]`）
- MLA 的 softmax scale 用 `attn_scale(nope+rope)` = 192，与 numpy 的
  `192^-0.5 * mscale^2 = 0.114721` 一致
- YaRN 的 inv_freq 用 rope 维 64（`ref_model.rs:454` / `paged_kv.rs:533`
  都传 `rope_hd`），有 `yarn_inv_freq_matches_hf` 测试覆盖
- shared experts 路径（SwiGLU 后加到 `h_acc`）与 numpy 一致
- MoE 路由权重语义（`norm_topk_prob=false` → `p * routed_scaling_factor`）
  与 config 一致
- chat 模板与 BOS：服务端 prompt_tokens=12，与 numpy 的 12 个 id 吻合
- checkpoint 字节级完好（4 分片 sha256 全匹配 HF etag，10 个小文件
  git blob sha1 全匹配）
- Q4 分组对齐：被 loader 切分/拼接的张量行长都是 32 的倍数
  （2048 / 512 / 1408），没有组跨越切分边界

## 未解决的困难

**这个 checkpoint 在这张卡上只有 Q4 路径能跑，拿不到未量化的基线**：

| 配置 | 显存需求 | 可用 |
|---|---|---|
| f16（无量化） | 30.56 GiB | 否（free 23.84 GiB） |
| FP8 | 59.82 GiB | 否 |
| Q4（无 Q4_DEVICE） | 30.56 GiB | 否 |
| Q4-on-device | 10.08 GiB | 是 |

所以无法在**全量 27 层**模型上做"量化 vs 未量化"的真机 A/B，只能在截断模型
上比。而截断模型（8 层）的 logit 分布极其平坦（top-20 只跨 4.4 个 logit），
对数值扰动过度敏感，不足以区分"实现有 bug"和"量化噪声被混沌放大"。

## 下一步建议

1. 直接为 `moe_grouped_gate_up_q4` / `moe_grouped_down_q4` 补单元测试：
   与 host 端 `Q4Tensor::dequantize()` + matmul 对比。这两个 kernel 目前
   **零测试覆盖**，是 #85 引入的，值得单独钉住。
2. 找一个能在 24GB 里以 f16 跑完的真实 MoE/MLA checkpoint（参数量 ≲ 8B），
   做全量模型的 f16 vs numpy 对拍。
3. 若以上都通过，则按"Q4 量化质量不足"记录为已知限制，并把首 token 翻转
   的幅度（1.44 → 0.67 的间距收窄）作为 Q4 的实测代价写进 roadmap。

## 复现

```powershell
# 服务端（Q4-on-device）
MACH_MODELS=.models MACH_MODEL=deepseek-v2-lite-chat \
  MACH_CONFIG=deepseek-v2-lite-chat/config.json \
  MACH_Q4=1 MACH_Q4_DEVICE=1 MACH_CAPACITY=4 MACH_MAX_SEQ=512 \
  ./target/release/mach-server.exe
```

```bash
# numpy 参考（含模拟 Q4）
FAKE_Q4=1 python .scratch/np_ref.py "The capital of France is" 8
```

脚本见 `.scratch/np_ref.py`（`FAKE_Q4` / `ROPE_LAYOUT` / `NLAY` / `REC_POS`）、
`.scratch/trunc.py`（构建 N 层截断 checkpoint）。

Refs #104

**Comment by gqf2008** [comment]
## 第一步结论：两个 grouped Q4 内核数值正确（直接单元证据）

按 #104 交接的建议，本批为 `moe_grouped_gate_up_q4` / `moe_grouped_down_q4` 补了直接单元测试（PR #109，已全绿）。

**覆盖的 expert-id 模式**（此前所有调用点只喂过连续 id，id 收集算术的错误恰好只能在未测过的模式下暴露）：

| 模式 | 说明 |
|---|---|
| 连续 `0..6` | 此前唯一被喂过的模式（基线） |
| 全行同一最高 id `[15;6]` | 同一专家被多行重复取用 |
| 跨池重复 `[15,0,15,7,15,3]` | 高 id + 混排重复 |
| batch=2 token-major | expert 15 同时出现在两个 token（row r 的 token 是 `r/topk`） |
| 真实行宽 d=2048/einter=1408 | gate_up 一行跨 1024 字节/64 组、down 一行 704 字节/44 组的对齐 |

**结果**：全部模式下与 spec 参考差 ≤1e-4。参考读者按文档布局**独立实现**（低半字节=偶元素、每 32 连续元素一组 scale，不调 `Q4Tensor::dequantize`——防读写共享同一约定 bug 两边一起错）；打包用真实 `Q4Tensor::quantize`，钉住 quantize → 裸上传 → 核内反量化整条接缝。

**嫌犯清单更新**——正式划掉：

- [x] `moe_grouped_gate_up_q4` / `moe_grouped_down_q4`（此前只有间接 A/B：Q4_DEVICE 开/关首 token 相同；现在是直接单元证据）

**剩余嫌犯**（下一批）：
1. 非专家张量的 Q4→**f16** 舍入：attention/dense MLP/shared experts/embed/lm_head 走反量化成 f16 再 GEMV，numpy 模拟是 f32；截断 8 层时 f16 路径落 numpy 次选（间距 0.18）属正常，但全 27 层 + Q4 误差复合可能放大
2. 路由/top-k/scatter 链路
3. 逐层分叉定位：全模型 f16（宿主反量化）vs numpy，找第一个分叉层

**Comment by gqf2008** [comment]
根因已定位并修复(PR #110,squash 11d03db):shared experts 守卫 !lw.shared_wg.is_null() 在全部 F16 上传路径(plain F16/Q4 存储/Q4-on-device/FP8)恒假,f32 指针在这些路径为 null(权重只活在 LayerDevF16 f16 表)。修复为 dtype 无关的 has_shared 标志。验证:DeepSeek-V2-Lite Q4-on-device 生成与 numpy 参考逐 token 一致(Paris 句),26 层路由全 SAME。同源遗留(model.rs 单序列 F16+MoE+shared、CPU offload 路径)另立 #111。
